Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -246,6 +246,39 @@ public final class SourceKitServer: LanguageServer {
 
     return nil
   }
+  
+  /// After the language service has crashed, send `DidOpenTextDocumentNotification`s to a newly instantiated language service for previously open documents.
+  func reopenDocuments(for languageService: ToolchainLanguageServer) {
+    queue.async {
+      guard let workspace = self.workspace else {
+        return
+      }
+      
+      for documentUri in workspace.documentManager.openDocuments where workspace.documentService[documentUri] === languageService {
+        guard let snapshot = workspace.documentManager.latestSnapshot(documentUri) else {
+          // The document has been closed since we retrieved its URI. We don't care about it anymore.
+          continue
+        }
+        
+        do {
+          // Close the document in the document manager so we can reopen it. If the document has been
+          // closed since, the try will fail and we won't try to reopen it.
+          try workspace.documentManager.close(documentUri)
+        } catch {
+          // The document was no longer open. Ignore it
+          continue
+        }
+        let textDocument = TextDocumentItem(uri: documentUri,
+                                            language: snapshot.document.language,
+                                            version: snapshot.version,
+                                            text: snapshot.text)
+        // Re-open the document but don't register for build system notifications again since
+        // the document is still open in the build system server from before the sourcektid crash.
+        self.openDocument(DidOpenTextDocumentNotification(textDocument: textDocument), workspace: workspace)
```

Comment:
> What's the reason for closing on the document manager and doing the full re-open rather than only doing `languageService.openDocument`?

Not 100% sure if this is correct (it's just been too longer) but failing tests show that it's something to do with build settings.

I believe, if we we just open the document in the language service again, we won't send the file's build settings over (of course we could wire that up as well), but I think it's easier to just close the file and then take the normal path of reopening the file again.

> Also, isn't this comment about "don't register for build system notifications" incorrect for the current implementation? It's going to re-register when reopening the document.

That was just a leftover from the old implementation that I should have removed.

Reply:
>  I believe, if we we just open the document in the language service again, we won't send the file's build settings over (of course we could wire that up as well), but I think it's easier to just close the file and then take the normal path of reopening the file again.

Ah okay.  Specifically the call to `workspace.buildSystemManager.registerForChangeNotifications` in `openDocument` will trigger sending the settings again.

I'm concerned that this approach entangles this function with implementation details of `openDocument` and how it interacts with the build system.  Currently, the document remains registered with the build system the whole time, and in fact this only works because double-registering for change notifications is idempotent.  If we changed registration to count the number of registrations, we might not send the settings again the second time, or this could "leak" the registration when we try to unregister later.

Now that I see the full picture, I'm leaning towards fully closing the document (e.g. call `closeDocument` or extract a method that both can call so that you don't need to construct a Notification wrapper). That will unregister for change notifications, keeping the calls balanced, and also clear `documentToPendingQueue` which seems like a good idea.  This will be slightly less optimal -- e.g. if the build system drops the cached settings and then we need to re-query them immediately -- but I think it's more maintainable.  Simplicity seems particularly important for code that is only exercised during crash recovery, since it will be used rarely and be hard to reproduce the specific timing of crashes.  What do you think?

Reply:
Sounds like a very good idea and a lot cleaner than the current approach of only closing it in the document manager. 

And I wouldn't be concerned about performance in crash scenarios. They shouldn't occur anyway and if they do, I'm totally fine with performance that’s equivalent to re-opening the workspace.